llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -4601,8 +4601,32 @@ VectorizationFactor LoopVectorizationPlanner::selectVectorizationFactor() {
         continue;
 
       InstructionCost C = CM.expectedCost(VF);
-      VectorizationFactor Candidate(VF, C, ScalarCost.ScalarCost);
 
+      // Add on other costs that are modelled in VPlan, but not in the legacy
+      // cost model.
+      VPCostContext CostCtx(CM.TTI, *CM.TLI, CM.Legal->getWidestInductionType(),
+                            CM, CM.CostKind);
+      if (VPRegionBlock *VectorRegion = P->getVectorLoopRegion()) {
+        for (VPBlockBase *Block :
+             vp_depth_first_shallow(VectorRegion->getEntry())) {
+          if (!isa<VPBasicBlock>(Block))
+            continue;
+          for (VPRecipeBase &R : *cast<VPBasicBlock>(Block)) {
+            if (auto *VPI = dyn_cast<VPInstruction>(&R)) {
```

Comment:
might be worth 

```suggestion
            auto *VPI = dyn_cast<VPInstruction>(&R);
            if (!VPI) 
              continue
```

to reduce the already high nesting level slightly

Reply:
Done